stunnel 5.61 support

julek-wolfssl opened the pull request:
- New/Implemented API
  - `SSL_has_pending`
  - `wolfSSL_CertManagerLoadCRLFile`
  - `wolfSSL_LoadCRLFile`
  - `wolfSSL_CTX_LoadCRLFile`
  - `wolfSSL_CTX_add_session`
- Access to session cache is now atomic
  - Adding and getting sessions to and from the local cache is now atomic.
  - The new internal `wolfSSL_GetSessionFromCache` requires a destination object to be supplied when retrieving from the cache so that items can be retrieved independently from the cache. For most existing calls, the destination is `ssl->session`.
  - `PREALLOC_SESSION_TICKET_LEN` defines how much memory is temporarily allocated for the ticket if it doesn't fit in the static session buffer.
  - Before this pull request, `wolfSSL_get_session` always returned a pointer to the internal session cache. The user can't tell if the underlying session hasn't changed before it calls `wolfSSL_set_session` on it. This PR adds a define `NO_SESSION_CACHE_REF` (for now only defined with `OPENSSL_COMPATIBLE_DEFAULTS`) that makes wolfSSL only return a pointer to `ssl->session`. The issue is that this makes the pointer returned non-persistent ie: it gets free'd with the `WOLFSSL` object. This pull request leverages the lightweight `ClientCache` to "increase" the size of the session cache. The hash of the session ID is checked to make sure that the underlying session hasn't changed.
- Calling chain certificate API (for example `wolfSSL_CTX_use_certificate_chain_file`) no longer requires an actual chain certificate PEM file to be passed in as input. `ProcessUserChain` error in `ProcessBuffer` is ignored if it returns that it didn't find a chain.
- Add `WOLFSSL_TICKET_HAVE_ID` macro. When defined tickets will include the original session ID that can be used to lookup the session in internal cache. This is useful for fetching information about the peer that doesn't get sent in a resumption (such as the peer's certificate chain).
  - Add `ssl->session->altSessionID` field because `ssl->session.sessionID` is used to return the "bogus" session ID sent by the client in TLS 1.3 and when using tickets in TLS <= 1.2
- `OPENSSL_COMPATIBLE_DEFAULTS` changes
  - Define `WOLFSSL_TRUST_PEER_CERT` and certificates added as CA's will also be loaded as trusted peer certificates
- Seperate `internalCacheOff` and `internalCacheLookupOff` options to govern session addition and lookup
- `VerifyServerSuite` now determines if RSA is available by checking for it directly and not assuming it as the default if static ECC is not available
- `WOLFSSL_SESSION` changes
  - `ssl->session->altSessionID` added to return a dynamic session when internalCacheOff is set
  - `ssl->session.refPtr` removed in favor of making `ssl->session` a seperate heap allocated object that can be returned to the user
- If `SSL_MODE_AUTO_RETRY` is set then retry should only occur during a handshake with a limit
- `WOLFSSL_TRUST_PEER_CERT` code now always uses `cert->subjectHash` for the `cm->tpTable` table row selection
- Change some error message names to line up with OpenSSL equivalents
- Run certificate setup callback before `MatchSuite`
- Refactor clearing `ASN_NO_PEM_HEADER` off the error queue into a macro
- `wolfSSL_get_peer_certificate` now returns a duplicated object meaning that the caller needs to free the returned object
- Align `wolfSSL_CRYPTO_set_mem_functions` callbacks with OpenSSL API
- `wolfSSL_d2i_PKCS12_bio` now consumes the input BIO. It now supports all supported BIO's instead of only memory BIO.
- stunnel specific
  - When allocating a dynamic session, always do `wolfSSL_SESSION_set_ex_data(session, 0, (void *)(-1)`. This is to mimic the new index callback set in `SSL_SESSION_get_ex_new_index`.
- Fix comment in `wolfSSL_AES_cbc_encrypt`
- Trusted peer certificate suite tests need to have CRL disabled since we don't have the issuer certificate in the CA store if the certificates are only added as trusted peer certificates.

julek-wolfssl committed:
stunnel 5.61 support

- New/Implemented API
  - `SSL_has_pending`
  - `wolfSSL_CertManagerLoadCRLFile`
  - `wolfSSL_LoadCRLFile`
  - `wolfSSL_CTX_LoadCRLFile`
  - `wolfSSL_CTX_add_session`
- Calling chain certificate API (for example `wolfSSL_CTX_use_certificate_chain_file`) no longer requires an actual chain certificate PEM file to be passed in as input. `ProcessUserChain` error in `ProcessBuffer` is ignored if it returns that it didn't find a chain.
- Add `WOLFSSL_TICKET_HAVE_ID` macro. When defined tickets will include the original session ID that can be used to lookup the session in internal cache. This is useful for fetching information about the peer that doesn't get sent in a resumption (such as the peer's certificate chain).
  - Add `ssl->ticketSessionID` field because `ssl->session.sessionID` is used to return the "bogus" session ID sent by the client in TLS 1.3
- `OPENSSL_COMPATIBLE_DEFAULTS` changes
  - Define `WOLFSSL_TRUST_PEER_CERT` and certificates added as CA's will also be loaded as trusted peer certificates
  - Define `WOLFSSL_TLS13_MIDDLEBOX_COMPAT`
- Seperate `internalCacheOff` and `internalCacheLookupOff` options to govern session addition and lookup
- `VerifyServerSuite` now determines if RSA is available by checking for it directly and not assuming it as the default if static ECC is not available
- `WOLFSSL_SESSION` changes
  - `ssl->extSession` added to return a dynamic session when internalCacheOff is set
  - `ssl->session.refPtr` made dynamic and gets free'd in `SSL_ResourceFree`
- If `SSL_MODE_AUTO_RETRY` is set then retry should only occur during a handshake
- `WOLFSSL_TRUST_PEER_CERT` code now always uses `cert->subjectHash` for the `cm->tpTable` table row selection
- Change some error message names to line up with OpenSSL equivalents
- Run `MatchSuite` again if certificate setup callback installed and successful
- Refactor clearing `ASN_NO_PEM_HEADER` off the error queue into a macro
- `wolfSSL_get_peer_certificate` now returns a duplicated object meaning that the caller needs to free the returned object
- Allign `wolfSSL_CRYPTO_set_mem_functions` callbacks with OpenSSL API
- `wolfSSL_d2i_PKCS12_bio` now consumes the input BIO. It now supports all supported BIO's instead of only memory BIO.
- stunnel specific
  - Always return a session object even if we don't have a session in cache. This allows stunnel to save information in the session external data that will be transfered to new connections if the session is reused
  - When allocating a dynamic session, always do `wolfSSL_SESSION_set_ex_data(session, 0, (void *)(-1)`. This is to mimic the new index callback set in `SSL_SESSION_get_ex_new_index`.
- Fix comment in `wolfSSL_AES_cbc_encrypt`
- Trusted peer certificate suite tests need to have CRL disabled since we don't have the issuer certificate in the CA store if the certificates are only added as trusted peer certificates.
tested

dgarske commented on src/ssl.c:
```diff
@@ -15468,24 +15565,22 @@ int wolfSSL_SetSession(WOLFSSL* ssl, WOLFSSL_SESSION* session)
 
     /* lock session cache row */
-    if (SESSION_ROW_LOCK(&SessionCache[row]) != 0) {
+    if (row != INVALID_SESSION_ROW &&
```
This code was specifically designed to avoid XMALLOC/XFREE during the session lock. Please consider these changes.

dgarske reviewed (changes requested):
Great work. A few minor items to review.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -23508,26 +23660,34 @@ WOLFSSL_SESSION* wolfSSL_GetSessionRef(WOLFSSL* ssl)
-    const word32     refSize = (word32)OFFSETOF(WOLFSSL_SESSION, refPtr) +
-                                       (word32)sizeof(wc_ptr_t);
-    int refCount = 0;
+    const word32     refSize = (word32)OFFSETOF(WOLFSSL_SESSION, _masterSecret);
```
Do you mean increased? This change makes it so that it will copy over all fields apart from the static ones. This includes the external data as well.

julek-wolfssl commented:
Also need to add new session fields to i2d and d2i api.

julek-wolfssl committed:
Access to session cache is now atomic

- Adding and getting sessions to and from the local cache is now atomic.
  - The new internal `wolfSSL_GetSessionFromCache` requires a destination object to be supplied when retrieving from the cache so that items can be retrieved independently from the cache. For most existing calls, the destination is `ssl->session`.
  -`PREALLOC_SESSION_TICKET_LEN` defines how much memory is temporarily allocated for the ticket if it doesn't fit in the static session buffer.

dgarske commented on src/ssl.c:
```diff
@@ -12459,7 +12460,7 @@ WOLFSSL_SESSION* wolfSSL_get_session(WOLFSSL* ssl)
     WOLFSSL_ENTER("SSL_get_session");
     if (ssl)
-        return wolfSSL_GetSession(ssl, NULL, 1);
+        return &ssl->session;
```
This doesn't work. `wolfSSL_get_session` is called... WOLFSSL object can be free'd between resumption.

julek-wolfssl replied:
That is correct because `wolfSSL_get_session` should return a reference to the working session (that is what `SSL_get_session` does). Users who want to retain a copy of the session should always use `wolfSSL_get1_session` as that will guarantee either an incremented refCount of a duplicated object.

dgarske replied:
This will break existing compatibility. Did you consider making the WOLFSSL_SESSION in WOLFSSL a pointer and allocating from heap? It would not increase memory use, since the WOLFSSL object is already allocated from the heap.

julek-wolfssl replied:
>This will break existing compatibility.

While true, the assumption is that a pointer directly to the cache entry is just as unsafe. The client user has no way of knowing if the session didn't change before setting it in a new SSL object. We could make `wolfSSL_get_session` and `wolfSSL_get1_session` do the same thing but that will cause leaks since `wolfSSL_get_session` is meant to return a reference to the SSL object's working session.

>Did you consider making the WOLFSSL_SESSION in WOLFSSL a pointer and allocating from heap?

This is an interesting idea. It should be fairly easy to do. The only issue would be that the heap is further fragmented. Do you think I should pursue this direction?

dgarske replied:
Yes please. You'll have to track that the allocated WOLFSSL_SESSION was retrieved and assume to not free it on `wolfSSL_free`. I think the session_ref stuff already does something similar, which could all be cleaned up and combined.

julek-wolfssl replied:
Do you mean that the session should point into the cache? I was thinking it would be allocated on wolfSSL_new and free'd on wolfSSL_free. A user could retain the object by either calling wolfSSL_get1_session or the up_ref API.

dgarske replied:
No, allocated.

julek-wolfssl replied:
Got it. I'll try to get it done tomorrow.

dgarske reviewed (changes requested):
I think there is great work in this PR, but we need to be very careful about how this will impact our customers... memory use, locking performance, API compatibility, etc...

Will you look into the resume test failures? Thanks

julek-wolfssl commented on src/ssl.c:
```diff
@@ -3356,7 +3356,7 @@ WOLFSSL_API int wolfSSL_set_SessionTicket(WOLFSSL* ssl, const byte* buf,
                           DYNAMIC_TYPE_SESSION_TICK);
                 }
-                ssl->session.ticket = (byte*)XMALLOC(bufSz, ssl->heap,
+                ssl->session.ticket = (byte*)XMALLOC(bufSz, ssl->session.heap,
```
They should be set to the same thing. I'll make it consistent for simplicity.

julek-wolfssl replied:
Fixed just now.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -23752,6 +23931,109 @@ void wolfSSL_SESSION_free(WOLFSSL_SESSION* session)
+
+    XFREE(copy, NULL, DYNAMIC_TYPE_SESSION);
+
+    SESSION_ROW_UNLOCK(sessRow);
```
This is addressed in the commit that makes access to the local cache atomic. There are no system calls during the cache access.

julek-wolfssl commented on wolfcrypt/src/error.c:
```diff
@@ -252,7 +252,11 @@ const char* wc_GetErrorString(int error)
+#ifndef OPENSSL_EXTRA
         return "ASN no signer error to confirm failure";
+#else
+        return "certificate verify failed";
```
Yup. Stunnel checks for log messages in their `make check` Python testing suite.

julek-wolfssl committed:
Allocate `ssl->session` separately on the heap

- Refactor session cache access into `AddSessionToCache` and `wolfSSL_GetSessionFromCache`

julek-wolfssl committed:
Fixes for Jenkins tests

- Move test to `HAVE_IO_TESTS_DEPENDENCIES`
- Implement `wolfSSL_trust_peer_cert`
- have{cipher} options weren't being set with only RSA enabled

julek-wolfssl committed:
Add backwards compatibility for `wolfSSL_get_session`

Before this pull request, `wolfSSL_get_session` always returned a pointer to the internal session cache. The user can't tell if the underlying session hasn't changed before it calls `wolfSSL_set_session` on it. This PR adds a define `NO_SESSION_CACHE_REF` (for now only defined with `OPENSSL_COMPATIBLE_DEFAULTS`) that makes wolfSSL only return a pointer to `ssl->session`. The issue is that this makes the pointer returned non-persistent ie: it gets free'd with the `WOLFSSL` object. This commit leverages the lightweight `ClientCache` to "increase" the size of the session cache. The hash of the session ID is checked to make sure that the underlying session hasn't changed.

dgarske commented on src/ssl.c:
```diff
@@ -15656,274 +15820,432 @@ static int get_locked_session_stats(word32* active, word32* total,
-        session->sessionIDSz = ID_LEN;
+    (void)error;
+
+    WOLFSSL_ENTER("AddSession");
```
Please add check for `ssl->session == NULL`.

dgarske reviewed (changes requested):
Great work on this! Lots to review here. First pass feedback.

julek-wolfssl commented on src/internal.c:
```diff
@@ -26712,41 +26712,41 @@ int SendCertificateVerify(WOLFSSL* ssl)
-        XFREE(ssl->session.ticket, ssl->heap, DYNAMIC_TYPE_SESSION_TICK);
-        ssl->session.ticket = ssl->session._staticTicket;
-        ssl->session.ticketLenAlloc = 0;
+    if (ssl->session->ticketLenAlloc > 0) {
```
The `ssl->session` member is allocated when the `WOLFSSL` object is created and `free`'ed only on `wolfSSL_free`. It is used extensively everywhere. Other members aren't tested for either like `ssl->arrays` or `ssl->suites`. I'm not sure if this is really necessary. Are you sure?

```
$ grep -r -P --include='*.[ch]' '\-\>session\-\>' | wc -l
233
```

julek-wolfssl committed:
Fix misc memory issues

- Make `InternalTicket` memory alignment independent

julek-wolfssl commented on src/internal.c:
```diff
@@ -30513,6 +30504,9 @@ static int DoSessionTicket(WOLFSSL* ssl, const byte* input, word32* inOutIdx,
     #endif
+#endif
+#ifdef WOLFSSL_TICKET_HAVE_ID
+        byte            id[ID_LEN];
```
So do you think this should be added permanently for all configurations with the session cache enabled?

julek-wolfssl commented on src/internal.c:
```diff
@@ -20201,7 +20193,11 @@ const char* wolfSSL_ERR_reason_error_string(unsigned long e)
         return "wrong client/server type";
 
     case NO_PEER_CERT :
+#ifndef OPENSSL_EXTRA
```
One fixed value with comment to explain compatibility.

julek-wolfssl commented on src/internal.c:
```diff
@@ -29955,6 +29942,15 @@ static int DoSessionTicket(WOLFSSL* ssl, const byte* input, word32* inOutIdx,
 
         ret = MatchSuite(ssl, &clSuites);
+#ifdef OPENSSL_EXTRA
+        /* Give user last chance to provide a cert for cipher selection */
```
I was thinking that the callback could be dependent on the cipher chosen but that doesn't make sense since the certificate selection will heavily influence the ciphersuite. Changing to one call after the cert callback.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -15656,274 +15820,432 @@ static int get_locked_session_stats(word32* active, word32* total,
+        if (error == 0) {
+            /* Lock row */
+            sessRow = &SessionCache[clientSession->serverRow];
+            error = SESSION_ROW_LOCK(sessRow);
```
`clientSession->serverRow` is checked above. Its an unsigned value so we only need to check if it greater than the maximum.

julek-wolfssl commented on configure.ac:
```diff
@@ -6822,6 +6823,15 @@ AC_ARG_ENABLE([optflags],
 AS_CASE(["$CFLAGS $CPPFLAGS"],[*'WOLFSSL_TRUST_PEER_CERT'*],[ENABLED_TRUSTED_PEER_CERT=yes])
 
 
+AS_CASE(["$CFLAGS $CPPFLAGS $AM_CFLAGS"],[*'OPENSSL_COMPATIBLE_DEFAULTS'*],
```
If its not part of the spec I'm removing it entirely. It won't be enabled for any port.

julek-wolfssl committed:
Peeking can't return a `WOLFSSL_ERROR_WANT_READ` in compatibility mode

dgarske reviewed (changes requested):
Lots of excellent work in this PR and looking forward to getting it merged. Just a few minor feedback items.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -15382,61 +15467,140 @@ static int SslSessionCacheOff(const WOLFSSL* ssl, const WOLFSSL_SESSION* session
+            XFREE(tmpTicket, output->heap, DYNAMIC_TYPE_TMP_BUFFER);
+#endif
+#endif
+        return WOLFSSL_FAILURE;
```
Github doesn't want to show more context on where you're commenting but it looks like this is the error case when `SESSION_ROW_LOCK` fails. No need to unlock the row if locking it failed.

julek-wolfssl committed:
Rebase and make `wolfSSL_CTX_up_ref` always available

`wolfSSL_CTX_up_ref` is a small and potentially useful API for users so it doesn't need to be restricted only to the compatibility layer. The reference counting mechanisms are always available anyway. This just exposes the functionality to the user.

julek-wolfssl commented on tests/api.c:
```diff
@@ -6199,6 +6229,191 @@ static void test_wolfSSL_get_finished(void)
+            test_wolfSSL_CTX_add_session_server_ctx == NULL) {
+        AssertNotNull(test_wolfSSL_CTX_add_session_server_ctx
+                = wolfSSL_get_SSL_CTX(ssl));
+        AssertIntEQ(wolfSSL_CTX_up_ref(wolfSSL_get_SSL_CTX(ssl)),
```
I made `wolfSSL_CTX_up_ref` always available. As I explain in the commit message:
>`wolfSSL_CTX_up_ref` is a small and potentially useful API for users so it doesn't need to be restricted only to the compatibility layer. The reference counting mechanisms are always available anyway. This just exposes the functionality to the user.

julek-wolfssl committed:
Prevent possibility of an infinite retry loop and resource exhaution

Reported in ZD13606

dgarske reviewed (approved):
Big changes! Session support massive refactor.
